src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyCodeFixProvider.cs

```diff
@@ -0,0 +1,111 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    internal abstract class AbstractMakeFieldReadonlyCodeFixProvider
+        <TSymbolSyntax, TFieldDeclarationSyntax>
+        : SyntaxEditorBasedCodeFixProvider
+        where TSymbolSyntax : SyntaxNode
+        where TFieldDeclarationSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics[0], c)),
+                context.Diagnostics);
+            return SpecializedTasks.EmptyTask;
+        }
+
+        private async Task FixWithEditorAsync(
+            Document document, SyntaxEditor editor, ImmutableArray<Diagnostic> diagnostics,
+            CancellationToken cancellationToken)
+        {
+            var declarators = new List<TSymbolSyntax>();
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+                declarators.Add(root.FindNode(diagnosticSpan).FirstAncestorOrSelf<TSymbolSyntax>());
+            }
+
+            MakeFieldReadonly(document, editor, declarators);
+        }
+
+        private async void MakeFieldReadonly(Document document, SyntaxEditor editor, List<TSymbolSyntax> declarators)
+        {
+            var declaratorsByField = declarators.GroupBy(g => g.FirstAncestorOrSelf<TFieldDeclarationSyntax>());
+
+            foreach (var fieldDeclarators in declaratorsByField)
+            {
+                var declarationDeclarators = GetVariableDeclarators(fieldDeclarators.Key);
+
+                if (declarationDeclarators.Count == fieldDeclarators.Count())
+                {
+                    editor.SetModifiers(fieldDeclarators.Key, editor.Generator.GetModifiers(fieldDeclarators.Key) | DeclarationModifiers.ReadOnly);
+                }
+                else
+                {
+                    var model = await document.GetSemanticModelAsync().ConfigureAwait(false);
+                    var generator = editor.Generator;
+                    
+                    foreach (var declarator in declarationDeclarators.Reverse())
+                    {
+                        var symbol = (IFieldSymbol)model.GetDeclaredSymbol(declarator);
+                        var modifiers = generator.GetModifiers(fieldDeclarators.Key);
+
+                        var newDeclaration = generator.FieldDeclaration(symbol.Name,
+                                                                        generator.TypeExpression(symbol.Type),
```

Comment:
i'm the fence if we should go back to the symbol here.  technically you can figure out hte name from the declarator.  And you can just use the same type node that was on the original field declaration right?